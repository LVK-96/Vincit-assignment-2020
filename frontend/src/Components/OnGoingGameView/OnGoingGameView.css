.onGoingGameContainer {
  height: 100%;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;
  margin: 20px;
}

.messageContainer {
  font-size: large;
  margin: 20px;
  height: 30px;
}

.gameContainer {
  display: flex;
  flex-direction: column;
}

.points {
  font-size: x-large;
  text-align: center;
  margin: 20px;
}

.playButton {
  background-color: red;
  color: white;
  font-size: xxx-large;
  padding: 20px;
}

.restartContainer {
  margin: 20px;
}

.restartButton {
  background-color: red;
  color: white;
  font-size: xxx-large;
  padding: 20px;
}
