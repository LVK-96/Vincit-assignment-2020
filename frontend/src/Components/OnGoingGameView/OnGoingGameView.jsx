import React, { useState } from 'react';
import gameService from '../../Services/game';
import usersService from '../../Services/users';
import Button from '../Button';
import Points from '../Points';
import GameMessage from '../GameMessage';
import './OnGoingGameView.css';

const MessageContainer = ({ loading, messageBackgroundColor, message }) => {
  const msg = message ? `Sait ${message.reward} pistett√§, seuraavaan voittoon ${message.untillNext} painallusta!` : '';
  if (!loading) {
    return (
      <div style={{ backgroundColor: messageBackgroundColor, color: 'white' }} className="messageContainer">
        <GameMessage message={msg} />
      </div>
    );
  }

  return (
    <div className="messageContainer">
      ...
    </div>
  );
};

const OnGoingGameView = ({
  setMessage, setPoints, message, points,
}) => {
  // Keep track of ongoing timeout so it can be canceled
  const [currentTimeOut, setCurrentTimeOut] = useState(null);
  // When loading, disable button and show loading icon instead of message
  const [loading, setLoading] = useState(false);
  const [messageBackgroundColor, setMessageBackGroundColor] = useState('white');

  const handlePlayClick = async () => {
    try {
      clearTimeout(currentTimeOut); // Stop old timeout so message doesn't get hidden too soon
      setLoading(true);

      const result = await gameService.play();
      setPoints(points - 1 + result.reward);
      setMessage(result);
      if (!result.reward) setMessageBackGroundColor('red');
      else setMessageBackGroundColor('green');

      setCurrentTimeOut(setTimeout(() => {
        setMessage(null);
        setMessageBackGroundColor('white');
      }, 5000)); // Message is visible for 5 seconds

      setLoading(false);
    } catch (e) {
      window.alert('Failed to play game!');
      setLoading(false);
      setMessageBackGroundColor('white');
    }
  };

  const handleRestartClick = async () => {
    try {
      setLoading(true);

      const user = await usersService.resetPoints();
      setMessage(null);
      setPoints(user.points);
      setMessageBackGroundColor('white');

      setLoading(false);
    } catch (e) {
      window.alert('Failed to restart game!');
      setLoading(false);
      setMessageBackGroundColor('white');
    }
  };

  let buttonText = 'Pelaa';
  let handleButtonClick = handlePlayClick;
  // Show correct text and use correct click handler when game needs to be restarted
  if (points < 1) {
    buttonText = 'Aloita alusta';
    handleButtonClick = handleRestartClick;
  }

  return (
    <>
      <MessageContainer
        message={message}
        loading={loading}
        messageBackgroundColor={messageBackgroundColor}
      />
      <div className="gameContainer">
        <div className="gameButtonContainer">
          <Button className="gameButton" text={buttonText} handleClick={handleButtonClick} disabled={loading} visible />
        </div>
        <div className="pointsContainer">
          <Points className="points" points={points} />
        </div>
      </div>
    </>
  );
};

export default OnGoingGameView;
