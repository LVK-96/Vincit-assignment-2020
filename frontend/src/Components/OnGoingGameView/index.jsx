import OnGoingGameView from './OnGoingGameView';

export default OnGoingGameView;
