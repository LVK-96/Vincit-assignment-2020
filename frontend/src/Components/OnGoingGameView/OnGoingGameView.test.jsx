import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { render, cleanup } from '@testing-library/react';
import renderer from 'react-test-renderer';
import OnGoingGameView from './OnGoingGameView';

jest.mock('../../Services/game.jsx');

afterEach(cleanup);

test('renders correctly', () => {
  const component = render(
    <OnGoingGameView />
  );

  expect(component.container).toHaveTextContent(
    'Pelaa'
  );

  expect(component.container).toHaveTextContent(
    'Pisteet: '
  );
});

test('snapshot test', () => {
  const component = renderer.create(
    <OnGoingGameView />
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
