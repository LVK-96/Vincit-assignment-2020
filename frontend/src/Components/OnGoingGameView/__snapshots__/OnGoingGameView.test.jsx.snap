// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
Array [
  <div
    className="message-container"
    style={
      Object {
        "backgroundColor": "white",
        "color": "white",
      }
    }
  >
    <div>
      
    </div>
  </div>,
  <div
    className="game-container"
  >
    <div
      className="game-button-container"
    >
      <button
        className="game-button"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Pelaa
      </button>
    </div>
    <div
      className="points-container"
    >
      <div
        className="points"
      >
        Pisteet: undefined
      </div>
    </div>
  </div>,
]
`;
